pkg/discovery/worker/dispatcher.go

```diff
@@ -0,0 +1,96 @@
+package worker
+
+import (
+	"fmt"
+	"math"
+
+	kubeClient "k8s.io/client-go/kubernetes"
+	api "k8s.io/client-go/pkg/api/v1"
+
+	"github.com/turbonomic/kubeturbo/pkg/discovery/probe"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/task"
+
+	"github.com/golang/glog"
+)
+
+type DispatcherConfig struct {
+	kubeClient  *kubeClient.Clientset
+	probeConfig *probe.ProbeConfig
+
+	workerCount int
+}
+
+func NewDispatcherConfig(kubeClient *kubeClient.Clientset, probeConfig *probe.ProbeConfig, workerCount int) *DispatcherConfig {
+	return &DispatcherConfig{
+		kubeClient:  kubeClient,
+		probeConfig: probeConfig,
+		workerCount: workerCount,
+	}
+}
+
+type Dispatcher struct {
+	config     *DispatcherConfig
+	workerPool chan chan *task.Task
+}
+
+func NewDispatcher(config *DispatcherConfig) *Dispatcher {
+	return &Dispatcher{
+		config: config,
+		// TODO use maxWorker count for now. Improve in the future once we find a good way to get the number of task.
+		// TODO If we allow worker number burst (# of workers > maxWorker), then the extra worker would block on registering. Or we use a threshold for burst number.
+		workerPool: make(chan chan *task.Task, config.workerCount),
+	}
+}
+
+func (d *Dispatcher) Init(c *ResultCollector) {
+	for i := 0; i < d.config.workerCount; i++ {
+		workerConfig := NewK8sDiscoveryWorkerConfig(d.config.kubeClient, d.config.probeConfig.StitchingPropertyType)
+		for _, mc := range d.config.probeConfig.MonitoringConfigs {
+			workerConfig.WithMonitoringWorkerConfig(mc)
+		}
+		// create workers
+		discoveryWorker, err := NewK8sDiscoveryWorker(workerConfig)
+		if err != nil {
+			fmt.Errorf("failed to build discovery worker %s", err)
+		}
+
+		go discoveryWorker.RegisterAndRun(d, c)
+	}
+}
+
+func (d *Dispatcher) WorkerPool() chan chan *task.Task {
+	return d.workerPool
+}
+
+func (d *Dispatcher) Dispatch(nodes []*api.Node, workerCount int) int {
+	// make sure when len(node) < workerCount, worker will receive at most 1 node to discover
+	perTaskNodeLength := int(math.Ceil(float64(len(nodes)) / float64(workerCount)))
```

Comment:
(1) A queue of tasks does not conflict with worker pool, and the architecture is simpler.
        https://play.golang.org/p/Y_sl7Ioeqs
   (dispatcher awareness of the consumer is not related to the architecture.)
  
(2)  Since each worker keeps live as long as the channel is not close, 
        the public info can be kept in the worker. 

(3) If multi-pods are needed for a cluster, then we should implement a distributed solution, and review it then.  And It will be similar amount of coding from current implementation to a distributed implementation, in either way.